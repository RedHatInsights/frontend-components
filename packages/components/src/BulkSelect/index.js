export { default } from './BulkSelect';
export { default as BulkSelect } from './BulkSelect';
