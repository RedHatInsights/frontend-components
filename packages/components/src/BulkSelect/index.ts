export { default } from './BulkSelect';
export { default as BulkSelect } from './BulkSelect';
export * from './BulkSelect';
