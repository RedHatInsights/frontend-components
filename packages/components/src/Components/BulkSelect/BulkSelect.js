import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Dropdown, DropdownItem, DropdownToggle, DropdownToggleCheckbox, Checkbox } from '@patternfly/react-core';

class BulkSelect extends Component {
    state = {
        isOpen: false,
        hasError: false
    }

    componentDidCatch = () => {
        console.error('Above error is caused because you are using outdated PF react core library. Count will not be \
visible unless you update it.');
        this.setState({ hasError: true });
    }

    onToggle = (isOpen) => {
        this.setState({
            isOpen
        });
    }

    render() {
        const { isOpen, hasError } = this.state;
        const { id, items, onSelect, checked, toggleProps, count, className, ...props } = this.props;

        return (
            <Fragment>
                { items && items.length > 0 ? <Dropdown
                    onSelect={ () => this.onToggle(false) }
                    { ...props }
                    className={ classnames(className, 'ins-c-bulk-select') }
                    toggle={ (
                        <DropdownToggle
                            { ...toggleProps }
                            splitButtonItems={ [
                                <Fragment key="split-checkbox">
                                    {
                                        hasError ? <DropdownToggleCheckbox
                                            id={ id ? `${id}-toggle-checkbox` : 'toggle-checkbox' }
                                            aria-label="Select all"
                                            onChange={ onSelect }
                                            checked={ checked || false }
                                        /> :
                                            <DropdownToggleCheckbox
                                                id={ id ? `${id}-toggle-checkbox` : 'toggle-checkbox' }
                                                aria-label="Select all"
                                                onChange={ onSelect }
                                                checked={ checked || false }
                                            >{ count ? `${count} selected` : '' }</DropdownToggleCheckbox>
                                    }
                                </Fragment>
                            ] }
                            onToggle={ this.onToggle }
                        />
                    ) }
                    isOpen={ isOpen }
                    dropdownItems={ [
                        ...count !== undefined && count > 0 ? [
                            <DropdownItem
                                key="count"
                                isDisabled
                                className={ !hasError ? 'ins-c-bulk-select__selected' : '' }
                            >
                                { count } Selected
                            </DropdownItem>
                        ] : [],
                        ...items.map((oneItem, key) => <DropdownItem
                            component="button"
                            key={ oneItem.key || key }
                            onClick={ (event) => oneItem.onClick && oneItem.onClick(event, oneItem, key) }
                        >
                            { oneItem.title }
                        </DropdownItem>)
                    ] }
                /> : <Checkbox
                    { ...props }
                    className={ classnames(className, 'ins-c-bulk-select') }
                    id={ `${id}-checkbox` }
                    isChecked={ checked }
                    onChange={ onSelect }
                /> }
            </Fragment>

        );
    }
}

BulkSelect.propTypes = {
    count: PropTypes.number,
    className: PropTypes.string,
    items: PropTypes.arrayOf(PropTypes.shape({
        title: PropTypes.string,
        onClick: PropTypes.func
    })),
    checked: PropTypes.bool,
    id: PropTypes.string,
    onSelect: PropTypes.func,
    toggleProps: PropTypes.any
};

BulkSelect.defaultProps = {
    className: '',
    count: 0,
    items: [],
    checked: false,
    onSelect: () => undefined
};

export default BulkSelect;
