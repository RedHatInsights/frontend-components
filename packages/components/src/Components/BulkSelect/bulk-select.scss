.ins-c-bulk-select {
    .pf-c-dropdown__toggle-check span.pf-c-dropdown__toggle-text {
        display: initial;
    }
    .ins-c-bulk-select__selected {
        display: none;
    }
}

@media only screen and (max-width: 768px) {
    .ins-c-bulk-select {
        .pf-c-dropdown__toggle-check span.pf-c-dropdown__toggle-text {
            display: none;
        }
        .ins-c-bulk-select__selected {
            display: initial;
        }
    }
}
