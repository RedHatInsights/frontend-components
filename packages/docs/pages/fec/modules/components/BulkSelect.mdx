
# BulkSelect

## Props

|name|type|default|description|
|----|----|-------|-----------|
|count|`number`|||
|className|`string`|`''`||
|items|BulkSelectItem[]|`[]`||
|checked|boolean&#124;null|`false`||
|id|`string`|||
|onSelect|DropdownToggleCheckboxProps['onChange']|`() => undefined`||
|toggleProps|<code>&#123;<br/><span className="default-prop-spacer"></span><span className="default-prop-spacer"></span><span className="default-prop-spacer"></span> "name": "DropdownToggleProps"<br />&#125;</code>|||
|isDisabled|<code>&#123;<br/><span className="default-prop-spacer"></span><span className="default-prop-spacer"></span><span className="default-prop-spacer"></span> "name": "boolean"<br />&#125;</code>|`false`||
|ouiaId|`string`|||
|ouiaSafe|<code>&#123;<br/><span className="default-prop-spacer"></span><span className="default-prop-spacer"></span><span className="default-prop-spacer"></span> "name": "boolean"<br />&#125;</code>|`true`||



