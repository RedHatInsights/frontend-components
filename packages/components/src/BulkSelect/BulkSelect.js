import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Dropdown, DropdownItem, DropdownToggle, DropdownToggleCheckbox, Checkbox } from '@patternfly/react-core';
import './bulk-select.scss';

class BulkSelect extends Component {
    state = {
        isOpen: false,
        hasError: false
    }

    componentDidCatch = () => {
        console.error('Above error is caused because you are using outdated PF react core library. Count will not be \
visible unless you update it.');
        this.setState({ hasError: true });
    }

    onToggle = (isOpen) => {
        this.setState({
            isOpen
        });
    }

    render() {
        const { isOpen, hasError } = this.state;
        const { id, isDisabled, items, onSelect, checked, toggleProps, count, className, ...props } = this.props;

        return (
            <Fragment>
                { items && items.length > 0 ? <Dropdown
                    onSelect={ () => this.onToggle(false) }
                    { ...props }
                    className={ classnames(className, 'ins-c-bulk-select') }
                    toggle={ (
                        <DropdownToggle
                            { ...toggleProps }
                            isDisabled={ isDisabled }
                            splitButtonItems={ [
                                <Fragment key="split-checkbox">
                                    {
                                        hasError ? <DropdownToggleCheckbox
                                            id={ id ? `${id}-toggle-checkbox` : 'toggle-checkbox' }
                                            aria-label="Select all"
                                            onChange={ onSelect }
                                            checked={ checked }
                                        /> :
                                            <DropdownToggleCheckbox
                                                id={ id ? `${id}-toggle-checkbox` : 'toggle-checkbox' }
                                                aria-label="Select all"
                                                onChange={ onSelect }
                                                isChecked={ checked }
                                            >{ count ? `${count} selected` : '' }</DropdownToggleCheckbox>
                                    }
                                </Fragment>
                            ] }
                            onToggle={ this.onToggle }
                        />
                    ) }
                    isOpen={ isOpen }
                    dropdownItems={ [
                        ...count !== undefined && count > 0 ? [
                            <DropdownItem
                                key="count"
                                isDisabled
                                className={ !hasError ? 'ins-c-bulk-select__selected' : '' }
                            >
                                { count } Selected
                            </DropdownItem>
                        ] : [],
                        ...items.map((oneItem, key) => <DropdownItem
                            component="button"
                            key={ oneItem.key || key }
                            onClick={ (event) => oneItem.onClick && oneItem.onClick(event, oneItem, key) }
                            { ...oneItem?.props }
                        >
                            { oneItem.title }
                        </DropdownItem>)
                    ] }
                /> : <Checkbox
                    { ...props }
                    className={ classnames(className, 'ins-c-bulk-select') }
                    id={ `${id}-checkbox` }
                    isChecked={ checked }
                    onChange={ onSelect }
                /> }
            </Fragment>

        );
    }
}

BulkSelect.propTypes = {
    count: PropTypes.number,
    className: PropTypes.string,
    items: PropTypes.arrayOf(PropTypes.shape({
        title: PropTypes.string,
        onClick: PropTypes.func
    })),
    checked: PropTypes.bool,
    id: PropTypes.string,
    onSelect: PropTypes.func,
    toggleProps: PropTypes.any
};

BulkSelect.defaultProps = {
    className: '',
    isDisabled: false,
    items: [],
    checked: false,
    onSelect: () => undefined
};

export default BulkSelect;
