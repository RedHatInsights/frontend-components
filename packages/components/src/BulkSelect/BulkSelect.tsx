import React, { Fragment, useRef, useState } from 'react';
import classnames from 'classnames';
import {
  Checkbox,
  Dropdown,
  DropdownItem,
  DropdownItemProps,
  DropdownToggle,
  DropdownToggleCheckbox,
  DropdownToggleCheckboxProps,
  DropdownToggleProps,
  getDefaultOUIAId,
} from '@patternfly/react-core';
import './bulk-select.scss';

export type BulkSelectItem = {
  key?: string | number;
  title?: string;
  props?: DropdownItemProps;
  onClick?: (
    event: MouseEvent | React.MouseEvent<any, MouseEvent> | React.KeyboardEvent<Element>,
    oneItem: BulkSelectItem,
    key: string | number
  ) => void;
};

export interface BulkSelectProps {
  count?: number;
  className?: string;
  items?: BulkSelectItem[];
  checked?: boolean | null;
  id?: string;
  onSelect?: DropdownToggleCheckboxProps['onChange'];
  toggleProps?: DropdownToggleProps;
  isDisabled?: boolean;
  ouiaId?: string;
  ouiaSafe?: boolean;
}

const BulkSelect: React.FunctionComponent<BulkSelectProps> = ({
  id,
  isDisabled = false,
  items = [],
  onSelect = () => undefined,
  checked = false,
  toggleProps,
  count,
  className = '',
  ouiaId,
  ouiaSafe = true,
  ...props
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const { current: hasError } = useRef(false);
  const { current: ouiaStateId } = useRef(getDefaultOUIAId('RHI/BulkSelect'));

  const onToggle = (isOpen: boolean) => setIsOpen(isOpen);

  const ouiaFinalId = ouiaId !== undefined ? ouiaId : ouiaStateId;

  return (
    <Fragment>
      {items && items.length > 0 ? (
        <Dropdown
          onSelect={() => onToggle(false)}
          {...props}
          className={classnames(className, 'ins-c-bulk-select')}
          ouiaId={ouiaFinalId}
          ouiaSafe={ouiaSafe}
          toggle={
            <DropdownToggle
              {...toggleProps}
              isDisabled={isDisabled}
              ouiaId={ouiaFinalId}
              splitButtonItems={[
                <Fragment key="split-checkbox">
                  {hasError ? (
                    <DropdownToggleCheckbox
                      id={id ? `${id}-toggle-checkbox` : 'toggle-checkbox'}
                      aria-label="Select all"
                      onChange={onSelect}
                      checked={checked}
                      ouiaId={ouiaFinalId}
                    />
                  ) : (
                    <DropdownToggleCheckbox
                      id={id ? `${id}-toggle-checkbox` : 'toggle-checkbox'}
                      aria-label="Select all"
                      onChange={onSelect}
                      isChecked={checked}
                      ouiaId={ouiaFinalId}
                    >
                      {count ? `${count} selected` : ''}
                    </DropdownToggleCheckbox>
                  )}
                </Fragment>,
              ]}
              onToggle={onToggle}
            />
          }
          isOpen={isOpen}
          dropdownItems={[
            ...(count !== undefined && count > 0
              ? [
                  <DropdownItem
                    key="count"
                    isDisabled
                    className={classnames({
                      'ins-c-bulk-select__selected': !hasError,
                    })}
                  >
                    {count} Selected
                  </DropdownItem>,
                ]
              : []),
            ...items.map((oneItem, key) => (
              <DropdownItem
                component="button"
                key={oneItem.key || key}
                ouiaId={`${ouiaFinalId}-${oneItem.key || key}`}
                onClick={(event) => oneItem.onClick && oneItem.onClick(event, oneItem, key)}
                {...oneItem?.props}
              >
                {oneItem.title}
              </DropdownItem>
            )),
          ]}
        />
      ) : (
        <Checkbox
          {...props}
          aria-label="Select all"
          className={classnames(className, 'ins-c-bulk-select')}
          id={`${id}-checkbox`}
          isChecked={checked}
          onChange={onSelect}
        />
      )}
    </Fragment>
  );
};

export default BulkSelect;
