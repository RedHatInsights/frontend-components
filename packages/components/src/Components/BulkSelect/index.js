export { default as BulkSelect } from './BulkSelect';
import './bulk-select.scss';
