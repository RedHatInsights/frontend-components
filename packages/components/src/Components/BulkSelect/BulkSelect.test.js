import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import BulkSelect from './BulkSelect';

describe('BulkSelect', () => {
    it('should render correctly - no data', () => {
        const wrapper = shallow(<BulkSelect />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render correctly', () => {
        const wrapper = shallow(<BulkSelect count={ 30 } items={ [
            {
                title: 'Select all',
                onClick: jest.fn()
            }
        ] } id="some-id" />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render correctly', () => {
        const wrapper = shallow(<BulkSelect items={ [
            {
                title: 'Select all',
                onClick: jest.fn()
            }
        ] } />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    describe('API', () => {
        it('should call on select with no items', () => {
            const onSelect = jest.fn();
            const wrapper = mount(<BulkSelect onSelect={ onSelect } />);
            wrapper.find('input[type="checkbox"]').first().simulate('change');
            expect(onSelect).toHaveBeenCalled();
        });

        it('should call on select', () => {
            const onSelect = jest.fn();
            const wrapper = mount(<BulkSelect items={ [
                {
                    title: 'Select all',
                    onClick: jest.fn()
                }
            ] } onSelect={ onSelect } />);
            wrapper.find('.pf-c-dropdown__toggle-check input[type="checkbox"]').first().simulate('change');
            expect(onSelect).toHaveBeenCalled();
        });

        it('should NOT call on select', () => {
            const onSelect = jest.fn();
            const wrapper = mount(<BulkSelect items={ [
                {
                    title: 'Select all',
                    onClick: jest.fn()
                }
            ] } />);
            wrapper.find('.pf-c-dropdown__toggle-check input[type="checkbox"]').first().simulate('change');
            expect(onSelect).not.toHaveBeenCalled();
        });

        it('should call first action', () => {
            const onSelect = jest.fn();
            const otherAction = jest.fn();
            const wrapper = mount(<BulkSelect items={ [
                {
                    title: 'Select all',
                    onClick: onSelect
                }, {
                    title: 'Some action',
                    onClick: otherAction
                }
            ] } />);
            wrapper.find('.pf-c-dropdown__toggle-button').first().simulate('click');
            wrapper.update();
            wrapper.find('.pf-c-dropdown__menu li .pf-c-dropdown__menu-item').first().simulate('click');
            expect(onSelect).toHaveBeenCalled();
            expect(otherAction).not.toHaveBeenCalled();
        });
    });
});
