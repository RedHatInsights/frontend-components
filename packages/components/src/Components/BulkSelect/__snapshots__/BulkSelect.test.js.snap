// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulkSelect should render correctly - no data 1`] = `
<Fragment>
  <Checkbox
    className="ins-c-bulk-select"
    id="undefined-checkbox"
    isChecked={false}
    isDisabled={false}
    isValid={true}
    onChange={[Function]}
  />
</Fragment>
`;

exports[`BulkSelect should render correctly 1`] = `
<Fragment>
  <Dropdown
    className="ins-c-bulk-select"
    dropdownItems={
      Array [
        <DropdownItem
          className="ins-c-bulk-select__selected"
          isDisabled={true}
        >
          30
           Selected
        </DropdownItem>,
        <DropdownItem
          component="button"
          onClick={[Function]}
        >
          Select all
        </DropdownItem>,
      ]
    }
    isOpen={false}
    onSelect={[Function]}
    toggle={
      <DropdownToggle
        onToggle={[Function]}
        splitButtonItems={
          Array [
            <React.Fragment>
              <DropdownToggleCheckbox
                aria-label="Select all"
                checked={false}
                className=""
                id="some-id-toggle-checkbox"
                isDisabled={false}
                isValid={true}
                onChange={[Function]}
              >
                30 selected
              </DropdownToggleCheckbox>
            </React.Fragment>,
          ]
        }
      />
    }
  />
</Fragment>
`;

exports[`BulkSelect should render correctly 2`] = `
<Fragment>
  <Dropdown
    className="ins-c-bulk-select"
    dropdownItems={
      Array [
        <DropdownItem
          component="button"
          onClick={[Function]}
        >
          Select all
        </DropdownItem>,
      ]
    }
    isOpen={false}
    onSelect={[Function]}
    toggle={
      <DropdownToggle
        onToggle={[Function]}
        splitButtonItems={
          Array [
            <React.Fragment>
              <DropdownToggleCheckbox
                aria-label="Select all"
                checked={false}
                className=""
                id="toggle-checkbox"
                isDisabled={false}
                isValid={true}
                onChange={[Function]}
              >
                
              </DropdownToggleCheckbox>
            </React.Fragment>,
          ]
        }
      />
    }
  />
</Fragment>
`;
